File: apps/mobile/modules/plaud-sdk/android/src/main/java/expo/modules/plaudsdk/PlaudWifiTransfer.kt
package expo.modules.plaudsdk

import android.os.Bundle
import android.os.Handler
import androidx.core.os.bundleOf
import expo.modules.kotlin.Promise
import java.io.File
import sdk.PlaudDeviceAgent
import sdk.audio.AudioExportFormat
import sdk.ble.wifi.IWifiTransferAgent
import sdk.ble.wifi.IWifiTransferAgent.WifiConnectionState
import sdk.ble.wifi.IWifiTransferAgent.WifiFileInfo
import sdk.ble.wifi.IWifiTransferAgent.BatchDownloadResult

/** Main-thread owner of one hotspot session. Credentials remain inside Plaud's SDK. */
internal class PlaudWifiTransfer(
  private val main: Handler,
  private val emit: (String, Bundle) -> Unit
) {
  private var opening: Promise? = null
  private var exporting: PlaudExportOperation? = null
  private var active = false
  private var joining = false
  private var epoch = 0L
  private var serial = ""
  private var files = emptySet<Long>()
  private var timeout: Runnable? = null
  val busy: Boolean get() = active || opening != null || exporting != null

  fun start(deviceSerial: String, promise: Promise) {
    if (PlaudExportGate.shared.isBusy()) {
      promise.reject("ERR_PLAUD_EXPORT_PENDING", "The previous transfer has not finished. Fully close and reopen the app.", null)
      return
    }
    if (busy || deviceSerial.isBlank() || !PlaudDeviceAgent.isConnected()) {
      promise.reject("ERR_PLAUD_WIFI_BUSY", "Connect the recorder and finish its previous transfer first.", null)
      return
    }
    val own = ++epoch
    serial = deviceSerial
    opening = promise
    active = true
    joining = false
    files = emptySet()
    timeout = Runnable {
      if (own == epoch) close("Wi-Fi connection timed out. Use Bluetooth or try again nearby.")
    }.also { main.postDelayed(it, 90_000) }
    try {
      PlaudDeviceAgent.setDeviceWiFi(true)
    } catch (_: Exception) {
      close("The recorder could not open Wi-Fi transfer.")
    }
  }

  fun hotspot(status: Int) {
    if (opening == null || joining) return
    if (status != 0) {
      close("This recorder could not open Wi-Fi transfer. Use Bluetooth or try again.")
      return
    }
    joining = true
    val own = epoch
    main.postDelayed({
      if (own == epoch && opening != null) {
        try {
          if (!PlaudDeviceAgent.startWifiTransfer(serial, callback(own))) {
            close("The phone could not start recorder Wi-Fi. Check Wi-Fi and nearby-device permissions.")
          }
        } catch (_: Exception) {
          close("The phone could not connect to recorder Wi-Fi.")
        }
      }
    }, 3_000)
  }

  fun export(sessionId: Long, directory: File, promise: Promise) {
    if (!active || opening != null || exporting != null || sessionId !in files ||
      PlaudDeviceAgent.getWifiAgent()?.getConnectionState() != WifiConnectionState.READY) {
      promise.reject("ERR_PLAUD_WIFI_NOT_READY", "The Wi-Fi recording session is not ready.", null)
      return
    }
    try {
      check(directory.isDirectory || directory.mkdirs())
    } catch (_: Exception) {
      promise.reject("ERR_PLAUD_WIFI_EXPORT", "Recording audio could not be saved.", null)
      return
    }
    val lease = PlaudExportGate.shared.acquire()
    if (lease == null) {
      promise.reject("ERR_PLAUD_EXPORT_PENDING", "The previous transfer has not finished. Fully close and reopen the app if it stopped responding.", null)
      return
    }
    val operation = PlaudExportOperation(main, promise, lease, sessionId, directory, "Wi-Fi", emit) { exporting = null }
    exporting = operation
    operation.started()
    try {
      PlaudDeviceAgent.exportAudioViaWiFi(sessionId, directory, AudioExportFormat.MP3, 1, operation)
    } catch (_: Exception) { operation.startFailed() }
  }

  fun close(message: String = "Wi-Fi transfer stopped.", stopSdk: Boolean = true) {
    exporting?.cancel()
    if (!active && opening == null) return
    ++epoch
    active = false
    joining = false
    timeout?.let { main.removeCallbacks(it) }
    timeout = null
    val pending = opening
    opening = null
    pending?.reject("ERR_PLAUD_WIFI", message, null)
    if (!stopSdk) return
    // Attempt every cleanup even if an SDK call throws. Keep export ownership until its
    // terminal callback; releasing it early would let a second exporter reuse SDK buffers.
    runCatching { PlaudDeviceAgent.getWifiAgent()?.stopWifiTransfer() }
    runCatching { PlaudDeviceAgent.setDeviceWiFi(false) }
    runCatching { PlaudDeviceAgent.endWiFiTransfer() }
  }

  private fun callback(own: Long) = object : IWifiTransferAgent.WifiTransferCallback {
    private fun accept(action: () -> Unit) {
      main.post {
        if (own == epoch && active) {
          try { action() } catch (_: Exception) { close("The recorder Wi-Fi session failed.") }
        }
      }
    }
    override fun onConnectionStateChanged(state: WifiConnectionState) = accept {
      when (state) {
        WifiConnectionState.READY -> {
          if (opening != null && PlaudDeviceAgent.getWifiAgent()?.getFileList() != true)
            close("The Wi-Fi recording list could not be requested.")
        }
        WifiConnectionState.ERROR, WifiConnectionState.DISCONNECTED ->
          close("The recorder Wi-Fi connection closed.")
        else -> Unit
      }
    }
    override fun onFileListReceived(files: List<WifiFileInfo>) = accept {
      val pending = opening ?: return@accept
      this@PlaudWifiTransfer.files = files.map { it.sessionId }.toSet()
      timeout?.let { main.removeCallbacks(it) }
      timeout = null
      opening = null
      pending.resolve(null)
    }
    override fun onDeviceBatteryUpdate(level: Int, charging: Boolean) = accept {
      emit("batteryState", bundleOf("batteryPercent" to level, "charging" to charging))
    }
    override fun onError(code: Int, message: String) = accept {
      close("The recorder reported a Wi-Fi error. Use Bluetooth or try again.")
    }
    override fun onWifiTransferStopped() = accept { close() }
    // Export progress/completion comes from AudioExporter, not the SDK's raw-file channel.
    override fun onHandshakeCompleted(info: String) {}
    override fun onTransferProgress(sessionId: Long, progress: Int, speed: Double) {}
    override fun onFileTransferCompleted(sessionId: Long, path: String) {}
    override fun onBatchDownloadStarted(total: Int) {}
    override fun onBatchDownloadProgress(current: Int, total: Int, filename: String) {}
    override fun onBatchDownloadCompleted(success: Int, failed: Int, results: List<BatchDownloadResult>) {}
    override fun onFileDeleteCompleted(success: Boolean, deletedCount: Int, error: String?) {}
  }
}

File: apps/mobile/modules/plaud-sdk/ios/PlaudWifiTransfer.swift
import ExpoModulesCore
import PlaudBleSDK
import PlaudDeviceBasicSDK

/// Main-queue owner of one Wi-Fi session and one exporter. Never replaces the BLE delegate.
final class PlaudWifiTransfer: NSObject, PlaudWiFiAgentProtocol, AudioExportCallback {
  // Native build configuration is authoritative, including for JavaScript updates.
  static var enabled: Bool {
    Bundle.main.object(forInfoDictionaryKey: "AptlyPlaudWifiTransferEnabled") as? Bool == true
  }
  private let emit: (String, [String: Any?]) -> Void
  private var opening: Promise?
  private var exporting: Promise?
  private var sessionId = 0
  private var scenes: [Int: Int] = [:]
  private var timeout: DispatchWorkItem?
  private var epoch = 0
  private var joining = false
  private(set) var active = false
  var busy: Bool { active || opening != nil || exporting != nil }

  init(emit: @escaping (String, [String: Any?]) -> Void) {
    self.emit = emit
    super.init()
  }

  func start(_ promise: Promise) {
    guard Self.enabled else {
      promise.reject("ERR_PLAUD_WIFI_DISABLED", "Wi-Fi transfer is unavailable in this build. Use Bluetooth.")
      return
    }
    guard !busy, PlaudDeviceAgent.shared.isConnected() else {
      promise.reject("ERR_PLAUD_WIFI_BUSY", "Connect the recorder and finish any previous transfer first.")
      return
    }
    epoch += 1
    let own = epoch
    opening = promise
    active = true
    joining = false
    scenes = [:]
    let wifi = PlaudWiFiAgent.shared
    wifi.delegate = self
    wifi.bleDevice = PlaudDeviceAgent.shared.recentConnectDevice
    let deadline = DispatchWorkItem { [weak self] in
      guard let self = self, own == self.epoch else { return }
      self.close("Wi-Fi connection timed out. Use Bluetooth or try again near the recorder.")
    }
    timeout = deadline
    DispatchQueue.main.asyncAfter(deadline: .now() + 90, execute: deadline)
    PlaudDeviceAgent.shared.setDeviceWiFi(open: true)
  }

  func hotspot(status: Int, ssid: String, password: String) {
    guard opening != nil, !joining else { return }
    guard status == 0, !ssid.isEmpty else {
      close("The recorder could not open Wi-Fi transfer. Use Bluetooth or try again.")
      return
    }
    joining = true
    let own = epoch
    // The device needs time to bring up its hotspot. Credentials stay inside native code.
    DispatchQueue.main.asyncAfter(deadline: .now() + 3) { [weak self] in
      guard let self = self, self.epoch == own, self.opening != nil else { return }
      PlaudWiFiAgent.shared.connectWifi(ssid, password, 60)
    }
  }

  func export(_ id: Int, promise: Promise) {
    guard Self.enabled else {
      promise.reject("ERR_PLAUD_WIFI_DISABLED", "Wi-Fi transfer is unavailable in this build. Use Bluetooth.")
      return
    }
    guard active, opening == nil, exporting == nil, scenes[id] != nil,
          PlaudWiFiAgent.shared.isConnected else {
      promise.reject("ERR_PLAUD_WIFI_NOT_READY", "The Wi-Fi recording session is not ready.")
      return
    }
    let directory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
      .appendingPathComponent("PlaudExports", isDirectory: true)
    do {
      try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    } catch {
      promise.reject("ERR_PLAUD_EXPORT", "The recording export folder could not be created.")
      return
    }
    sessionId = id
    exporting = promise
    PlaudWiFiAgent.shared.exportAudioViaWiFi(
      sessionId: id, outputDir: directory.path, format: .mp3, channels: 1, callback: self
    )
  }

  func close(_ message: String = "Wi-Fi transfer stopped.") {
    guard active || opening != nil else { return }
    epoch += 1
    active = false
    joining = false
    timeout?.cancel()
    timeout = nil
    let pending = opening
    opening = nil
    pending?.reject("ERR_PLAUD_WIFI", message)
    if exporting != nil {
      PlaudWiFiAgent.shared.stopSyncFile(sessionId, scenes[sessionId] ?? 1)
    }
    PlaudDeviceAgent.shared.setDeviceWiFi(open: false)
    PlaudDeviceAgent.shared.endWiFiTransfer()
    PlaudWiFiAgent.shared.disconnect()
    // Retain exporter ownership until its terminal callback, even after cancellation.
    // Starting another native export early can mix SDK audio buffers.
  }

  func wifiHandshake(_ status: Int) {
    DispatchQueue.main.async { [weak self] in
      guard let self = self, self.opening != nil, self.joining else { return }
      guard status == 0 else { self.close("The recorder Wi-Fi handshake failed."); return }
      PlaudWiFiAgent.shared.getFileList(Int(Date().timeIntervalSince1970), 0, false)
    }
  }

  func wifiFileList(_ files: [BleFile]) {
    DispatchQueue.main.async { [weak self] in
      guard let self = self, let promise = self.opening else { return }
      self.scenes = files.reduce(into: [:]) { result, file in result[file.sessionId] = file.scenes }
      self.timeout?.cancel()
      self.timeout = nil
      self.opening = nil
      promise.resolve(nil)
    }
  }

  func wifiFileListFail(_ status: Int) {
    DispatchQueue.main.async { [weak self] in self?.close("The Wi-Fi recording list could not be read.") }
  }
  func wifiClientFail() {
    DispatchQueue.main.async { [weak self] in self?.close("The phone could not connect to recorder Wi-Fi.") }
  }
  func wifiClose(_ status: Int) {
    DispatchQueue.main.async { [weak self] in self?.close("The recorder Wi-Fi connection closed.") }
  }
  func wifiCommonErr(_ cmd: Int, _ status: Int) {
    DispatchQueue.main.async { [weak self] in self?.close("The recorder reported a Wi-Fi error.") }
  }

  func onProgress(_ progress: Int, message: String) {
    DispatchQueue.main.async { [weak self] in
      guard let self = self, self.exporting != nil, self.active else { return }
      self.emit("exportProgress", ["sessionId": self.sessionId, "progress": progress, "message": "Receiving over Wi-Fi"])
    }
  }
  func onComplete(outputPath: String) {
    DispatchQueue.main.async { [weak self] in
      guard let self = self, let promise = self.exporting else { return }
      self.exporting = nil
      promise.resolve(["sessionId": self.sessionId, "outputPath": outputPath])
    }
  }
  func onError(_ error: String) {
    DispatchQueue.main.async { [weak self] in
      guard let self = self, let promise = self.exporting else { return }
      self.exporting = nil
      promise.reject("ERR_PLAUD_WIFI_EXPORT", "Wi-Fi audio transfer failed. Use Bluetooth or reconnect and try again.")
    }
  }
}
